concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -517,47 +507,54 @@ doReceiveTransaction tr slot = unlessShutDown $ do
                -- but this should be revised to add a "accountExists" function in the future
                senderExists <- flip getAccount (transactionSender tx) =<< queryBlockState =<< lastFinalizedBlock
                if isNothing senderExists then return ResultNonexistingSenderAccount
-               else snd <$> doReceiveTransactionInternal tr slot
-        _ -> snd <$> doReceiveTransactionInternal tr slot
+               else snd <$> doReceiveTransactionInternal tr (utcTimeToTimestamp now) slot
+        _ -> snd <$> doReceiveTransactionInternal tr (utcTimeToTimestamp now) slot
     when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
     return ur
-
-    where
-          isExpiryTooLate now = do
-            maxTimeToExpiry <- rpMaxTimeToExpiry <$> getRuntimeParameters
-            let expiry = msgExpiry tr
-            return $ expiry > maxTimeToExpiry + utcTimeToTransactionTime now
+  where
+    isExpiryTooLate tx ts = do
+        maxTimeToExpiry <- rpMaxTimeToExpiry <$> getRuntimeParameters
+        let expiry = msgExpiry tx
+        return $ expiry > maxTimeToExpiry + utcTimeToTransactionTime ts
 
 -- |Add a transaction to the transaction table.  The 'Slot' should be
 -- the slot number of the block that the transaction was received with.
 -- This function should only be called when a transaction is received as part of a block.
 -- The difference from the above function is that this function returns an already existing
--- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
+-- transaction ipn case of a duplicate, ensuring more sharing of transaction data.
+-- This function also verifies the transactions incoming and adds them to the internal
+-- transaction verification cache such that it can be used by the 'Scheduler'.
+doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Timestamp -> Slot -> m (Maybe BlockItem, UpdateResult)
+doReceiveTransactionInternal tr ts slot = do
+   -- Don't accept the transaction if it was expired
+   if transactionExpired (msgExpiry tr) ts then return (Nothing, ResultStale)
+   else do
+    cache <- getTransactionVerificationCache   
+    (verRes, cache') <- runReaderT (verifyWithCache ts tr cache) =<< blockState =<< getFocusBlock
+    if not (isCacheable verRes) then do return (Nothing, mapTransactionVerificationResult verRes)
```

Comment:
For clarity it would be good to have a function that is more explicit. Here we are rejecting the entire block based on this result. So having something like
```
definitelyNotValid :: Result -> Bool
```
would be clearer.